### PR TITLE
Madrys et al's attack.

### PR DESCRIPTION
**Pull request by fartashf**
`mnist_tutorial_madryetal.py` is supposed to reproduce the results of the paper: https://arxiv.org/pdf/1706.06083.pdf

It gets better results.

Here are the first 10 and last 10 iterations of this command:
`python mnist_tutorial_madryetal.py --clean_train False --nb_epochs 100`

[INFO 2017-09-22 01:10:07,728 cleverhans] Epoch 0 took 227.005951166 seconds
Test accuracy on legitimate examples: 0.9814
Test accuracy on adversarial examples: 0.2328
[INFO 2017-09-22 01:14:29,324 cleverhans] Epoch 1 took 225.296789885 seconds
Test accuracy on legitimate examples: 0.9875
Test accuracy on adversarial examples: 0.4358
[INFO 2017-09-22 01:18:50,357 cleverhans] Epoch 2 took 224.643723011 seconds
Test accuracy on legitimate examples: 0.9879
Test accuracy on adversarial examples: 0.6959
[INFO 2017-09-22 01:23:12,155 cleverhans] Epoch 3 took 225.241991997 seconds
Test accuracy on legitimate examples: 0.9886
Test accuracy on adversarial examples: 0.7668
[INFO 2017-09-22 01:27:34,159 cleverhans] Epoch 4 took 225.454128027 seconds
Test accuracy on legitimate examples: 0.9901
Test accuracy on adversarial examples: 0.7875
[INFO 2017-09-22 01:31:55,593 cleverhans] Epoch 5 took 225.021591902 seconds
Test accuracy on legitimate examples: 0.9913
Test accuracy on adversarial examples: 0.8104
[INFO 2017-09-22 01:36:17,346 cleverhans] Epoch 6 took 225.302556038 seconds
Test accuracy on legitimate examples: 0.9920
Test accuracy on adversarial examples: 0.8072
[INFO 2017-09-22 01:40:38,634 cleverhans] Epoch 7 took 224.783905029 seconds
Test accuracy on legitimate examples: 0.9916
Test accuracy on adversarial examples: 0.8097
[INFO 2017-09-22 01:45:00,257 cleverhans] Epoch 8 took 225.015685797 seconds
Test accuracy on legitimate examples: 0.9922
Test accuracy on adversarial examples: 0.8171
[INFO 2017-09-22 01:49:21,331 cleverhans] Epoch 9 took 224.659055948 seconds
Test accuracy on legitimate examples: 0.9915
Test accuracy on adversarial examples: 0.8321
[INFO 2017-09-22 01:53:42,967 cleverhans] Epoch 10 took 225.024608135 seconds
Test accuracy on legitimate examples: 0.9924
Test accuracy on adversarial examples: 0.8631

[INFO 2017-09-22 07:33:05,518 cleverhans] Epoch 90 took 223.041524172 seconds
Test accuracy on legitimate examples: 0.9905
Test accuracy on adversarial examples: 0.9433
[INFO 2017-09-22 07:37:25,188 cleverhans] Epoch 91 took 223.353102922 seconds
Test accuracy on legitimate examples: 0.9904
Test accuracy on adversarial examples: 0.9438
[INFO 2017-09-22 07:41:43,964 cleverhans] Epoch 92 took 222.522644043 seconds
Test accuracy on legitimate examples: 0.9909
Test accuracy on adversarial examples: 0.9476
[INFO 2017-09-22 07:46:03,226 cleverhans] Epoch 93 took 223.163653135 seconds
Test accuracy on legitimate examples: 0.9901
Test accuracy on adversarial examples: 0.9436
[INFO 2017-09-22 07:50:21,975 cleverhans] Epoch 94 took 222.51416111 seconds
Test accuracy on legitimate examples: 0.9902
Test accuracy on adversarial examples: 0.9441
[INFO 2017-09-22 07:54:41,040 cleverhans] Epoch 95 took 222.864598989 seconds
Test accuracy on legitimate examples: 0.9910
Test accuracy on adversarial examples: 0.9451
[INFO 2017-09-22 07:58:59,766 cleverhans] Epoch 96 took 222.39167881 seconds
Test accuracy on legitimate examples: 0.9903
Test accuracy on adversarial examples: 0.9409
[INFO 2017-09-22 08:03:18,902 cleverhans] Epoch 97 took 222.829951048 seconds
Test accuracy on legitimate examples: 0.9900
Test accuracy on adversarial examples: 0.9391
[INFO 2017-09-22 08:07:37,987 cleverhans] Epoch 98 took 222.753010988 seconds
Test accuracy on legitimate examples: 0.9907
Test accuracy on adversarial examples: 0.9433
[INFO 2017-09-22 08:11:57,235 cleverhans] Epoch 99 took 223.244225979 seconds
Test accuracy on legitimate examples: 0.9903
Test accuracy on adversarial examples: 0.9418
[INFO 2017-09-22 08:12:33,438 cleverhans] Completed model training.

### CODE REVIEW

File: cleverhans_tutorials/tutorial_models.py [hunk 1]
```diff
@@ -87,13 +94,16 @@ def set_input_shape(self, input_shape):
                                                    self.output_channels)
         assert len(kernel_shape) == 4
         assert all(isinstance(e, int) for e in kernel_shape), kernel_shape
-        init = tf.random_normal(kernel_shape, dtype=tf.float32)
-        init = init / tf.sqrt(1e-7 + tf.reduce_sum(tf.square(init),
-                                                   axis=(0, 1, 2)))
+        if self.weight_init == 't_normal':
```
Comment (AlexeyKurakin) [inline cleverhans_tutorials/tutorial_models.py]: Duplicated code here and in linear layer. It would be nice if you can get rid of code duplication.

File: cleverhans_tutorials/tutorial_models.py [hunk 2]
```diff
@@ -77,7 +83,8 @@ def fprop(self, x):
 
 class Conv2D(Layer):
 
-    def __init__(self, output_channels, kernel_shape, strides, padding):
+    def __init__(self, output_channels, kernel_shape, strides, padding,
+                 bias_init=0., weight_init=None):
```
Comment (AlexeyKurakin) [inline cleverhans_tutorials/tutorial_models.py]: bias_init is not used

File: cleverhans/attacks.py [hunk 3]
```diff
@@ -948,3 +948,161 @@ def jsma(sess, x, predictions, grads, sample, target, theta, gamma=np.inf,
                     clip_min, clip_max)
     elif back == 'th':
         raise NotImplementedError("Theano jsma not implemented.")
+
+
+class MadryEtAl(Attack):
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans/attacks.py]: This was suggested by @goodfeli, as many attacks that already exist do some sort of projected gradient descent.

File: cleverhans/attacks.py [hunk 4]
```diff
@@ -7,7 +7,7 @@
 import cleverhans.utils as utils
 from cleverhans.model import Model, CallableModelWrapper
 
-import logging
+# import logging
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans/attacks.py]: Removed.

File: cleverhans/attacks.py [hunk 5]
```diff
@@ -948,3 +948,161 @@ def jsma(sess, x, predictions, grads, sample, target, theta, gamma=np.inf,
...
+
+        labels, nb_classes = self.get_or_guess_labels(x, kwargs)
+        self.targeted = self.y_target is not None
+
+        # Initialize loop variables
+        adv_x = self.attack(x)
+
+        return adv_x
+
+    def parse_params(self, eps=0.3, eps_iter=0.01, nb_iter=10, y=None,
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans/attacks.py]: Changed to 40 to make it similar to MNIST experiments of the paper.

File: cleverhans/attacks.py [hunk 6]
```diff
@@ -948,3 +948,161 @@ def jsma(sess, x, predictions, grads, sample, target, theta, gamma=np.inf,
...
+        if self.ord not in [np.inf, 1, 2]:
+            raise ValueError("Norm order must be either np.inf, 1, or 2.")
+        if self.back == 'th':
+            error_string = ("ProjectedGradientDescentMethod is"
+                            " not implemented in Theano")
+            raise NotImplementedError(error_string)
+
+        return True
+
+    def clip_eta(self, eta):
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans/attacks.py]: Done.

File: cleverhans/attacks.py [hunk 7]
```diff
@@ -948,3 +948,161 @@ def jsma(sess, x, predictions, grads, sample, target, theta, gamma=np.inf,
...
+        loss = model_loss(y, preds)
+        if self.targeted:
+            loss = -loss
+        grad, = tf.gradients(loss, adv_x)
+        scaled_signed_grad = self.eps_iter * tf.sign(grad)
+        adv_x = adv_x + scaled_signed_grad
+        adv_x = tf.clip_by_value(adv_x, self.clip_min, self.clip_max)
+        eta = adv_x - x
+        eta = self.clip_eta(eta)
+        return x, eta, y
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans/attacks.py]: fixed.

File: cleverhans/attacks_tf.py [hunk 8]
```diff
@@ -8,7 +8,7 @@
 from six.moves import xrange
 import tensorflow as tf
 import warnings
-import logging
+# import logging
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans/attacks_tf.py]: removed.

File: cleverhans_tutorials/mnist_tutorial_madryetal.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
+"""
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans_tutorials/mnist_tutorial_madryetal.py]: @npapernot suggested that in order to keep the tutorials as intended, we don't duplicate the tutorials nor we add flags. He suggested to put such things in the examples, where we don't run tests. If this is agreed on, I can fully remove this file.

[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans_tutorials/mnist_tutorial_madryetal.py]: I decided to remove this tutorial. I don't think there is a need for it.

File: cleverhans/attacks.py [hunk 10]
```diff
@@ -948,3 +948,161 @@ def jsma(sess, x, predictions, grads, sample, target, theta, gamma=np.inf,
...
+            y = y / tf.reduce_sum(y, 1, keep_dims=True)
+        y = tf.stop_gradient(y)
+
+        for i in range(self.nb_iter):
+            x, eta, y = self.attack_single_step(x, eta, y)
+
+        adv_x = x + eta
+        if self.clip_min is not None and self.clip_max is not None:
+            adv_x = tf.clip_by_value(adv_x, self.clip_min, self.clip_max)
+        adv_x = tf.stop_gradient(adv_x)
```
[the remark this answers is not in the document]
Comment (fartashf) [reply cleverhans/attacks.py]: fixed.

File: cleverhans/attacks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexeyKurakin) [reply cleverhans/attacks.py]: Sounds good.

**Comment by fartashf** [comment]
Great. Thank you.

**Comment by npapernot** [comment]
Thanks for the PR @fartashf and for reviewing @AlexeyKurakin
